src/raft.c

```diff
@@ -115,19 +115,185 @@ void entryAttachRaftReq(RedisRaftCtx *rr, raft_entry_t *entry, RaftReq *req)
+
+        for (int j = 0; j < num_keys; j++) {
+            found += RedisModule_KeyExists(rr->ctx, cmd->argv[keyindex[j]]);
+        }
+        RedisModule_Free(keyindex);
+
+        /* shortcut as we know the result is now SomeExist */
+        if (found != 0 && found != total_keys) {
+            return SomeExist;
+        }
+    }
+
+    if (found != total_keys) {
+        return (found == 0) ? NoneExist : SomeExist;
+    }
+
+    return (total_keys == 0) ? NoKeys : AllExist;
+}
+
+/* figure out the "owner shardgroup"
+ *
+ * the owner shardgroup is either
+ *
+ * 1) a shardgroup that owns the slot as a stable slot
+ * 2) a shardgroup that owns the slot as a migrating slot.  by definition is this shardgroup is local, then it
+ *    can't also be importing (a single RedisRaft cluster cannot be both importing and migrating the same slot
+ * 3) a shardgroup marked as local (i.e. corresponding to this cluster) that owns the slot as an importing slot and
+ *    has a RaftRedisCommandArray marked as asking
+ */
+
+ShardGroup * GetCommandShardGroup(RedisRaftCtx *rr, RaftRedisCommandArray *cmds)
+{
+    ShardGroup *sg = rr->sharding_info->stable_slots_map[cmds->slot];
+    if (sg != NULL) {
+        return sg;
+    }
+
+    sg = rr->sharding_info->migrating_slots_map[cmds->slot];
+    if (sg && sg->local) {
+        return sg;
+    }
+
+    if (cmds->asking) {
+        ShardGroup *isg = rr->sharding_info->importing_slots_map[cmds->slot];
+        if (isg && isg->local) {
+            return isg;
+        }
+    }
+
+    return sg;
+}
+
+RRStatus validateRaftRedisCommandArray(RedisRaftCtx *rr, RedisModuleCtx *reply_ctx, RaftRedisCommandArray *cmds) {
+    /* Make sure hash slot is mapped and handled locally. */
+    ShardGroup *sg = GetCommandShardGroup(rr, cmds);
+    if (!sg) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "CLUSTERDOWN Hash slot is not served");
+        }
+        return RR_ERROR;
+    }
+
+     if (cmds->slot < 0) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "CLUSTERERROR should not have a negative slot here");
+        }
+        return RR_ERROR;
+    }
+    unsigned int slot = (unsigned int) cmds->slot;
+
+    SlotRangeType slot_type = SLOTRANGE_TYPE_UNDEF;
+    for (size_t i = 0; i < sg->slot_ranges_num; i++) {
+        if (sg->slot_ranges[i].start_slot <= slot && slot <= sg->slot_ranges[i].end_slot) {
+            slot_type = sg->slot_ranges[i].type;
+            break;
+        }
+    }
+
+    if (slot_type == SLOTRANGE_TYPE_UNDEF) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "ERR internal error, couldn't associate a shardgroup slot to this request");
+        }
+        return RR_ERROR;
+    }
+
+    if (!sg->local) {
+        if (reply_ctx) {
+            if (sg->next_redir >= sg->nodes_num) {
+                sg->next_redir = 0;
+            }
+
+            replyRedirect(reply_ctx, cmds->slot, &sg->nodes[0].addr);
```

Comment:
either remove the next_redir code above or replace 0 in sg->nodes[0].addr with the next_redir value

Reply:
@yossigo thoughts about this?

Reply:
@sjpotter I commented about this topic in a different location.